services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/AbstractPushMonitor.java

```diff
@@ -659,10 +670,57 @@ protected void onPartitionStatusChange(OfflinePushStatus offlinePushStatus) {
     checkWhetherToStartBufferReplayForHybrid(offlinePushStatus);
   }
 
+  protected DisableReplicaCallback getDisableReplicaCallback(String kafkaTopic) {
+    DisableReplicaCallback callback = new DisableReplicaCallback() {
+      private final Map<String, Set<Integer>> disabledReplicaMap = new HashMap<>();
+      private final Set<String> enabledReplicas = new HashSet<>();
+
+      @Override
+      public void disableReplica(String instance, int partitionId) {
+        LOGGER.warn(
+            "Disabling errored out leader replica of {} partition: {} on host {}",
+            kafkaTopic,
+            partitionId,
+            instance);
+        helixAdminClient.enablePartition(
+            false,
+            clusterName,
+            instance,
+            kafkaTopic,
+            Collections.singletonList(HelixUtils.getPartitionName(kafkaTopic, partitionId)));
+        disabledReplicaMap.computeIfAbsent(instance, k -> new HashSet<>()).add(partitionId);
+        enabledReplicas.remove(instance);
+      }
+
+      @Override
+      public boolean isReplicaDisabled(String instance, int partitionId) {
+        Set<Integer> disabledPartitions = disabledReplicaMap.getOrDefault(instance, null);
+        if (disabledPartitions != null) {
+          return disabledPartitions.contains(partitionId);
+        }
+        if (enabledReplicas.contains(instance)) {
+          return false;
+        }
+        helixClientThrottler.maybeThrottle(1);
+        Map<String, List<String>> helixMap = helixAdminClient.getDisabledPartitionsMap(clusterName, instance);
+        if (helixMap.containsKey(kafkaTopic)) {
+          disabledPartitions = helixMap.get(kafkaTopic).stream().map(Integer::parseInt).collect(Collectors.toSet());
+          disabledReplicaMap.computeIfAbsent(instance, k -> new HashSet<>()).addAll(disabledPartitions);
+          return disabledPartitions.contains(partitionId);
+        } else {
+          enabledReplicas.add(instance);
```

Comment:
The naming and logic is a little tricky, so can we put an emptySet for the good instance in `disabledReplicaMap`?

Reply:
The logic could become more consistent:
1. `computeIfAbsent` to get all the disabled replicas for a certain instance.
2. check whether the partition id is in the returned set or not.